File: Iteration1CSE3063Proje/src/main/java/com/cse3063/rag/impl/TemplateAnswerAgent.java
package com.cse3063.rag.impl;

import com.cse3063.rag.interfaces.AnswerAgent;
import com.cse3063.rag.interfaces.Retriever.Hit;
import java.util.Collections;
import java.util.List;

public class TemplateAnswerAgent implements AnswerAgent {

    @Override
    public Answer answer(List<String> queryTerms, List<Hit> topHits) {
        if (topHits.isEmpty()) {
            return new Answer("I couldn't find any information about that.", Collections.emptyList());
        }

        Hit bestHit = topHits.get(0);
        String bestSentence = findBestSentence(bestHit.text, queryTerms);

        String answerText = String.format("Your answer: %s. See: %s:%s",
                bestSentence, bestHit.docId, bestHit.chunkId);

        return new Answer(answerText, Collections.singletonList(bestHit.docId + ":" + bestHit.chunkId));
    }

    private String findBestSentence(String text, List<String> terms) {
        String[] sentences = text.split("[.!?]");
        String best = sentences[0];
        int maxMatches = 0;

        for (String sentence : sentences) {
            int matches = 0;
            for (String term : terms) {
                if (sentence.toLowerCase().contains(term)) {
                    matches++;
                }
            }
            if (matches > maxMatches) {
                maxMatches = matches;
                best = sentence;
            }
        }
        return best.trim();
    }
}

File: python_port/rag/impl/template_answer_agent.py
import re
from typing import List
from ..interfaces import AnswerAgent, Hit, Answer

class TemplateAnswerAgent(AnswerAgent):
    def answer(self, query_terms: List[str], top_hits: List[Hit]) -> Answer:
        if not top_hits:
            return Answer("I couldn't find any information about that.", [])
        
        best_hit = top_hits[0]
        best_sentence = self.find_best_sentence(best_hit.text, query_terms)
        
        answer_text = f"Your answer: {best_sentence}. See: {best_hit.docId}:{best_hit.chunkId}"
        
        return Answer(answer_text, [f"{best_hit.docId}:{best_hit.chunkId}"])

    def find_best_sentence(self, text: str, terms: List[str]) -> str:
        sentences = re.split(r'[.!?]', text)
        # Filter out empty strings from split
        sentences = [s.strip() for s in sentences if s.strip()]
        
        if not sentences:
            return text # Fallback if no sentence splitting happened

        best = sentences[0]
        max_matches = 0
        
        for sentence in sentences:
            matches = 0
            lower_sentence = sentence.lower()
            for term in terms:
                if term in lower_sentence:
                    matches += 1
            
            if matches > max_matches:
                max_matches = matches
                best = sentence
                
        return best
